### PR TITLE
[Storage] [Typing] [File Datalake] Decoupled `_data_lake_service_client.py` and async

### PR DESCRIPTION
**Pull request by weirongw23-msft**
Decoupled `_data_lake_service_client.py` and async. Will defer resolving `mypy` errors until all clients have been decoupled.

Edit: Will need to order the imports and let CI pass :)

Edit 2: Imports *should* be ordered.

### CODE REVIEW

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_data_lake_service_client_helper.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# -------------------------------------------------------------------------
```
Comment (weirongw23-msft) [inline sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_data_lake_service_client_helper.py:1]: Theoretically should be `_data_lake_service_client_helpers.py` with an `s`.

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_data_lake_service_client_helper.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+# pylint: disable=docstring-keyword-should-match-keyword-only
+
+from typing import TYPE_CHECKING
+from urllib.parse import urlparse
+
+if TYPE_CHECKING:
+    from urllib.parse import ParseResult
+
+
+def _parse_url(account_url: str) -> "ParseResult":
```
Comment (weirongw23-msft) [inline sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_data_lake_service_client_helper.py:15]: Move this to `_parser` or merge with the next PR

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_data_lake_service_client_async.py [hunk 3]
```diff
@@ -86,44 +92,100 @@ class DataLakeServiceClient(AsyncStorageAccountHostsMixin, DataLakeServiceClient
...
+        :type credential:
+            ~azure.core.credentials.AzureNamedKeyCredential or
+            ~azure.core.credentials.AzureSasCredential or
+            ~azure.core.credentials.TokenCredential or
+            str or Dict[str, str] or None
+        :keyword str audience: The audience to use when requesting tokens for Azure Active Directory
+            authentication. Only has an effect when credential is of type TokenCredential. The value could be
+            https://storage.azure.com/ (default) or https://<account>.blob.core.windows.net.
+        :returns: A DataLakeServiceClient.
+        :rtype: ~azure.storage.filedatalake.DataLakeServiceClient
```
Comment (weirongw23-msft) [inline sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_data_lake_service_client_async.py:178]: Do we need a sample for async datalake service client here? There's an identical one for the sync version, how about linking that one instead? Or we could leave as is :)

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_data_lake_service_client.py [hunk 4]
```diff
@@ -5,26 +5,36 @@
 # --------------------------------------------------------------------------
```
Comment (weirongw23-msft) [inline sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_data_lake_service_client.py:5]: Look for any lingering `dict(str, str)`

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_data_lake_service_client_helper.py (hunk 2 above)
Comment (vincenttran-msft) [reply sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_data_lake_service_client_helper.py:15]: What does this comment mean?

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_data_lake_service_client_async.py (hunk 3 above)
Comment (vincenttran-msft) [reply sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_data_lake_service_client_async.py:178]: After looking at what we did in Blob:
https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/storage/azure-storage-blob/azure/storage/blob/_blob_service_client.py#L175
https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/storage/azure-storage-blob/azure/storage/blob/aio/_blob_service_client_async.py#L179

And Queue:
https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/storage/azure-storage-queue/azure/storage/queue/_queue_service_client.py#L149
https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/storage/azure-storage-queue/azure/storage/queue/aio/_queue_service_client_async.py#L141

We just re-used the sync sample. I think this is fine and we definitely shouldn't just omit the sample just because we don't have a specific "async" example (which would just involve importing the async equiv. and maybe an `await` here and there 😆 

tl;dr add the sample and it's fine to reuse the sync sample

**Review by vincenttran-msft — APPROVED** [review]
LGTM just some small nits

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_data_lake_service_client_helper.py (hunk 1 above)
Comment (jalauzon-msft) [reply sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_data_lake_service_client_helper.py:1]: Not just theoretical :)

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_data_lake_service_client_helper.py (hunk 2 above)
Comment (jalauzon-msft) [reply sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_data_lake_service_client_helper.py:15]: Is the exact same function going to be used by multiple clients or is it different per client? If its the same for multiple clients, moving it to parsers makes sense. As for which PR, whatever works for you.

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_data_lake_service_client_async.py [hunk 5]
```diff
@@ -86,44 +92,100 @@ class DataLakeServiceClient(AsyncStorageAccountHostsMixin, DataLakeServiceClient
...
+
         self._client = AzureDataLakeStorageRESTAPI(self.url, base_url=self.url, pipeline=self._pipeline)
-        self._client._config.version = get_api_version(kwargs)  #pylint: disable=protected-access
+        self._client._config.version = get_api_version(kwargs)  # pylint: disable=protected-access
         self._loop = kwargs.get('loop', None)
 
-    async def __aenter__(self):
+    async def __aenter__(self) -> Self:
         await self._blob_service_client.__aenter__()
         return self
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_data_lake_service_client_async.py:130]: Unrelated to this PR but... Interesting that `aenter` does not call super but `aexit` does? The super here is now `AsyncStorageAccountHostsMixin` so it may make sense to call the super?

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_data_lake_service_client_async.py (hunk 3 above)
Comment (jalauzon-msft) [reply sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_data_lake_service_client_async.py:178]: Personally, I don't really think we need the sample if we don't have one already. This is a pretty easy thing to use. But if you want to add one to match other packages, I won't stop you :D

Comment (weirongw23-msft) [reply sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_data_lake_service_client_async.py:178]: Sure that sounds good!

Comment (weirongw23-msft) [reply sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_data_lake_service_client_async.py:178]: Mainly just like our packages to be as symmetrical as possible for a good UX.

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_data_lake_service_client_helper.py (hunk 1 above)
Comment (weirongw23-msft) [reply sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_data_lake_service_client_helper.py:1]: I also don't love the idea of having just one function in the helpers. I didn't see any other need to factor out functionality in the service client, and same with file system client as well. Should we get rid of the helpers file and say it's okay for some code duplication?

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_data_lake_service_client.py (hunk 4 above)
Comment (weirongw23-msft) [reply sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_data_lake_service_client.py:5]: Done

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_data_lake_service_client_async.py (hunk 3 above)
Comment (weirongw23-msft) [reply sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_data_lake_service_client_async.py:178]: The sample already exists, just gotta add to the docs after the decoupling work!

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_data_lake_service_client_async.py (hunk 5 above)
Comment (weirongw23-msft) [reply sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_data_lake_service_client_async.py:131]: Added the super call before `await self._blob_service_client.__aenter__()`

**Review by vincenttran-msft — APPROVED** [review]
LGTM! BTW you have `StreamConsumedError`s and misc. recording errors. Can you check to see if those are legitimate issues or if those are just leftover failures from `HttpResponse` changes, and that this branch is stale (aka you might need to merge `main` into `feature/storage-datalake-typing` and propagate it to these branches).

If you are confident these are not real test/recording failures and will be fixed with a merge from `main`, you could also force merge these into `feature/storage-datalake-typing` and then merge `main` into that branch and just resolve it all at once. Dealer's choice 😄

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_data_lake_service_client_async.py [hunk 6]
```diff
@@ -86,44 +92,110 @@ class DataLakeServiceClient(AsyncStorageAccountHostsMixin, DataLakeServiceClient
...
+            The credentials with which to authenticate. This is optional if the
+            account URL already has a SAS token, or the connection string already has shared
+            access key values. The value can be a SAS token string,
+            an instance of a AzureSasCredential from azure.core.credentials, an account shared access
+            key, or an instance of a TokenCredentials class from azure.identity.
+            Credentials provided here will take precedence over those in the connection string.
+        :type credential:
+            ~azure.core.credentials.AzureNamedKeyCredential or
+            ~azure.core.credentials.AzureSasCredential or
+            ~azure.core.credentials.TokenCredential or
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_data_lake_service_client_async.py:173]: Can we point this to the async version and the other necessary edits (i.e. L168 refers to sync). Apply this feedback to any other instances if applicable :)

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_data_lake_service_client_async.py [hunk 7]
```diff
@@ -189,7 +263,7 @@ def list_file_systems(self, name_starts_with=None,  # type: Optional[str]
             Flag specifying that system filesystems should be included.
             .. versionadded:: 12.6.0
         :returns: An iterable (auto-paging) of FileSystemProperties.
-        :rtype: ~azure.core.paging.ItemPaged[~azure.storage.filedatalake.FileSystemProperties]
+        :rtype: ~azure.core.async_paging.AsyncItemPaged[~azure.storage.filedatalake.FileSystemProperties]
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_data_lake_service_client_async.py:266]: Good catch here!

**Review by vincenttran-msft — APPROVED** [review]
Nit on the docstring, LGTM otherwise! :shipit:

**Review by jalauzon-msft — APPROVED** [review]
